### DISCUSSION TITLE
Operators/methods with by-ref vs by-val overloads resolution

### DISCUSSION THREAD

**Discussion by BladeWise** [dotnet/csharplang#7159] [Language Ideas]
I have noticed that having operator overloads that differ by by-ref/by-val parameters, makes the by-ref overload unusable, because there is no way to explicitly select it (e.g., you cannot use something like `in x + in y`.

For methods, omitting the `in` modifier and passing a by-ref will result in the by-val overload being selected by default.

See the following example:
```cs
    private delegate T ByRefOperator<T>(in T value) where T : unmanaged;
    
    [StructLayout(LayoutKind.Sequential)]
    public readonly struct MyVector2<T> where T : unmanaged, INumberBase<T>
    {
        public static MyVector2<T> operator -(MyVector2<T> vector) => new(T.One - vector.X, T.One - vector.Y);
        public static MyVector2<T> operator -(in MyVector2<T> vector) => new(T.One - vector.X, T.One - vector.Y);
        public static MyVector2<T> Negate(MyVector2<T> vector) => new(T.One - vector.X, T.One - vector.Y);
        public static MyVector2<T> Negate(in MyVector2<T> vector) => new(T.One - vector.X, T.One - vector.Y);
                
        public static MyVector2<double> TestOperator()
        {
            var v = new MyVector2<double>(1, 2);
            ref readonly MyVector2<double> vv = ref v;
            var y = -vv;
            return y;
        }
                
        public static MyVector2<double> TestMethod()
        {
            var v = new MyVector2<double>(1, 2);
            ref readonly MyVector2<double> vv = ref v;
            var y = MyVector2<double>.Negate(vv);
            return y;
        }
        
        public T X { get; }
        public T Y { get; }
        
        public MyVector2(T x, T y)
        {
            X = x;
            Y = y;
        }
    }
```
In both `TestOperator` and `TestMethod` the by-val overload is used, but while it is possible to forcefully select by-ref method overload (e.g., `MyVector2<double>.Negate(in vv)`), there is no way to do the same for the by-ref operator overload.

Shouldn't be possible to have an overload resolution strategy that takes into account the `in` modifier, so that even by-ref operator overloads can be selected?

**Comment by CyrusNajmabadi** [comment]
What's the use case for wanting both operators?

**Reply by BladeWise** [reply]
Generalized vector math, as an example, and more generally, generic structs that depending on the size of T could benefit from by-ref parameters instead of by-val ones, and vice-versa.
Providing both operator overloads means that the developer using the code can opt-in for by-ref operator overloads, the same way he can do for by-ref method overloads.

On top of that, the fact that the language allows the definition of both operator overloads, but provides no way to explicitly select one of them (while the compiler does not provide any warning about operator overload being not usable), can be confusing in my opinion.

**Reply by CyrusNajmabadi** [reply]
But if you have the by-ref operator, why would someone want the by-val one?  When is it appropriate for the caller to call one over the other?

**Reply by BladeWise** [reply]
I posed the same question in the C# Discord server, and the answer is that by-ref parameters are not a guarantee of performance in every scenario. The general consensus is that structs larger that 16 bytes should see some improvements by by-ref parameters, but it could depend on the optimizer and the target platform (it's ABI-dependant).
In my opinion, in such scenarios there could be merit in providing both overloads, and let the developer use the one that fits the specific scenario better.

**Reply by CyrusNajmabadi** [reply]
I would suggest just using normal methods then.  I htink this would def be confusing (and def strange sytnax) to have to have this in line with operator calls.

**Reply by BladeWise** [reply]
I do not suggest to provide a specific syntax for such overloads (I agree that it would be confusing).
My request is to ensure that writing this
```cs
        public static MyVector2<double> TestOperator()
        {
            var v = new MyVector2<double>(1, 2);
            ref readonly MyVector2<double> vv = ref v;
            var y = -vv;
            return y;
        }
```
will result in this overload being called
```cs
        public static MyVector2<T> operator -(in MyVector2<T> vector) => new(T.One - vector.X, T.One - vector.Y);
```
instead of
```cs
        public static MyVector2<T> operator -(MyVector2<T> vector) => new(T.One - vector.X, T.One - vector.Y);
```
which is what happens now.

Basically, my proposal is to take into account the by-ref/by-val nature of the parameter as part of the automatic overload resolution strategy.

**Reply by HaloFour** [reply]
Sometimes it'd be nice if you could call the method directly:

```cs
var y = MyVector<double>.op_Subtraction(in x, in y);
```

**Reply by BladeWise** [reply]
Yes, but would not make sense in this context if you want to provide operator-like syntax. 😞

**Reply by TahirAhmadov** [reply]
@BladeWise 
> will result in this overload being called

Sounds like you want the language to prefer the `in` overload for the operator, but I don't see how this makes things better. If there is a requirement that by-val is better in some cases, preferring the `in` overload will mean the by-val overload will never be used? I think the solution here (other than the weird `in` syntax in expressions) would be to simply choose which parameter strategy is expected to be used more often, and making that the operator, and making the other strategy to be the static method. Basically, if by-ref (`in`) is more often preferable, make that the operator and make the method by-val, and the user can decide what to do that way.
PS. Scratch that - I see what you mean - the `vv` is of type `ref`, and yes I agree in that case the `in` overload should be preferred.

**Reply by BladeWise** [reply]
Exactly, given that the language has now support for ref/ref readonly, using this information as part of the operator overload selection seems more natural than the current behavior.
